library/src/layout/par.rs

```diff
@@ -1233,13 +1272,30 @@ fn commit(
         }
     }
 
-    // Determine how much to justify each space.
+    // Determine how much addtional space is needed.
     let fr = line.fr();
-    let mut justification = Abs::zero();
-    if remaining < Abs::zero() || (line.justify && fr.is_zero()) {
+    let mut justification_ratio = 0.0;
+    let mut extra_justification = Abs::zero();
+
+    let shrink = line.shrinkability();
+    let stretch = line.stretchability();
+    if remaining < Abs::zero() && shrink > Abs::zero() {
+        // Attempt to reduce the length of the line, using shrinkability.
+        let new_remaining = Abs::min(remaining + shrink, Abs::zero());
+        justification_ratio = (remaining - new_remaining) / shrink;
```

Comment:
This should be equivalent to the following, right?
```
justification_ratio = (remaining / shrink).max(-1.0);
```
Which is (at least to me), a lot easier to understand.

Reply:
Fixed in f9dc4b2ce177968b5f670299c561a7855c50ff1b